pkgs/by-name/cu/cups-kyocera-3500-4500/package.nix

```diff
@@ -22,18 +22,29 @@ stdenv.mkDerivation rec {
 
   dontStrip = true;
 
-  src = fetchzip {
+  src = fetchurl {
     # Steps to find the release download URL:
     # 1. Go to https://www.kyoceradocumentsolutions.us/en/support/downloads.html
     # 2. Search for printer model, e.g. "TASKalfa 6053ci"
     # 3. Locate e.g. "Linux Print Driver (9.3)" in the list
     urls = [
-      "https://dam.kyoceradocumentsolutions.com/content/dam/gdam_dc/dc_global/executables/driver/product_085/KyoceraLinuxPackages-${date}.tar.gz"
+      "https://www.kyoceradocumentsolutions.us/content/download-center-americas/us/drivers/drivers/MA_PA_4500ci_Linux_gz.download.gz"
+      "https://web.archive.org/web/20241123173620/https://www.kyoceradocumentsolutions.us/content/download-center-americas/us/drivers/drivers/MA_PA_4500ci_Linux_gz.download.gz"
```

Comment:
Wouldn't it be better to put the permalink first? (Having a mutable URL first feels like a small regression to me.)

Reply:
I think having the canonical url first is better, since it is more "trustworthy" ? if we have the permalink first would never fallback to the "correct" one?

Reply:
The package is a fixed-output derivation, so I'm fairly comfortable about trust here. And I think describing it as a "regression" is a bit misleading, given the previous URL was on the vendor's site too. But particularly for a proprietary driver package, I'd rather use the vendor's bandwidth rather than the Wayback Machine's, where possible, not to mention the fact that I expect the vendor's site is significantly faster.

FWIW, I had a look at a selection of packages that use the Wayback Machine as a source; they all either only use the Wayback Machine or they use the Wayback Machine as a fallback URL if the first URL fails.

Reply:
> The package is a fixed-output derivation, so I'm fairly comfortable about trust here.

Agreed.

> And I think describing it as a "regression" is a bit misleading, given the previous URL was on the vendor's site too. 

The point wasn't about whether or not the vendor site was first, but about permalink vs mutable URL.

> But particularly for a proprietary driver package, I'd rather use the vendor's bandwidth rather than the Wayback Machine's, where possible, not to mention the fact that I expect the vendor's site is significantly faster.

Sure, that's a good argument. But personally I'd rather have long-term reliable Nix expressions than short term fast ones. I expect the mutable URL to change on the next driver update which invalidates the Nix expression. (I think fetchurl takes the first URL that downloads successfully; it doesn't try until one has the expected hash. So the 2nd fallback URL won't be used as long as the first URL provides _something_.)

> FWIW, I had a look at a selection of packages that use the Wayback Machine as a source; they all either only use the Wayback Machine or they use the Wayback Machine as a fallback URL if the first URL fails.

The ones that put Wayback Machine last, is the first URL a permalink or not? (Actually, I don't think that matters; you'll have to convince me that having a mutable URL first is a good idea from first principles, not by how many times a (possibly bad pattern) exists in nixpkgs already :-)

Reply:
Right, I'd misunderstood your concern. By "mutable URL" you mean you're concerned the vendor might release an updated version of this file at the same URL, which would lead to the build breaking, because fetchurl will only try later URLs if earlier ones actually fail to download, rather than successfully downloading a file that ultimately leads to a hash failure.

That makes sense to me, and is a risk I hadn't considered. I suspect it's more likely that if Kyocera release a new package, they'll change the URL even though this one doesn't encode package versions or similar, but I absolutely understand not wanting to take the chance. I'll respin a new PR with the order reversed and a comment to explain what's going on.

Reply:
New PR at #358959

Reply:
@me-and: Thanks!